File: sdk/go/tracelens/span.go
package tracelens

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type Span struct {
	config    SpanConfig
	attrs     map[string]interface{}
	start     time.Time
	ended     bool
	Data      *SpanData
}

func NewSpan(config SpanConfig) *Span {
	return &Span{
		config: config,
		attrs:  make(map[string]interface{}),
		start:  time.Now(),
	}
}

func (s *Span) SetAttribute(key string, value interface{}) {
	if s.ended {
		return
	}
	s.attrs[key] = value
}

func (s *Span) SetAttributes(attrs map[string]interface{}) {
	if s.ended {
		return
	}
	for k, v := range attrs {
		s.attrs[k] = v
	}
}

func (s *Span) End(cfg EndConfig) (*SpanData, error) {
	if s.ended {
		return nil, fmt.Errorf("span already ended")
	}
	s.ended = true
	latency := time.Since(s.start).Milliseconds()

	llmModel, _ := s.attrs["llm.model"].(string)
	inputTokens, _ := s.attrs["llm.input_tokens"].(int)
	if inputTokens == 0 {
		inputTokens = 0
	}
	outputTokens, _ := s.attrs["llm.output_tokens"].(int)
	costUSD, _ := s.attrs["cost.usd"].(float64)

	s.Data = &SpanData{
		SpanID:       uuid.New().String(),
		ParentID:     s.config.ParentID,
		AgentType:    s.config.Agent,
		ToolName:     s.config.Task,
		LLMModel:     &llmModel,
		InputTokens:  inputTokens,
		OutputTokens: outputTokens,
		LatencyMs:    latency,
		Status:       cfg.Status,
		CostUSD:      costUSD,
		Timestamp:    time.Now().UTC().Format(time.RFC3339Nano),
		Attributes:   s.attrs,
	}

	if cfg.ErrorMessage != nil {
		s.Data.ErrorMessage = cfg.ErrorMessage
	}

	return s.Data, nil
}

func (s *Span) IsEnded() bool {
	return s.ended
}

File: sdk/python/tracelens/span.py
import time
import uuid
from typing import Optional, Dict, Any
from .types import SpanConfig, SpanData


class Span:
    def __init__(self, config: SpanConfig):
        self.config = config
        self.attrs: Dict[str, Any] = {}
        self.start_time = time.time()
        self._ended = False
        self.data: Optional[SpanData] = None

    def set_attribute(self, key: str, value: Any) -> None:
        if self._ended:
            return
        self.attrs[key] = value

    def set_attributes(self, attrs: Dict[str, Any]) -> None:
        if self._ended:
            return
        self.attrs.update(attrs)

    def end(self, status: str = 'ok', error_message: Optional[str] = None) -> SpanData:
        if self._ended:
            raise RuntimeError('Span already ended')
        self._ended = True
        latency = int((time.time() - self.start_time) * 1000)

        self.data = SpanData(
            span_id=str(uuid.uuid4()),
            parent_id=self.config.parent_id,
            agent_type=self.config.agent,
            tool_name=self.config.task,
            llm_model=self.attrs.get('llm.model'),
            input_tokens=self.attrs.get('llm.input_tokens', 0),
            output_tokens=self.attrs.get('llm.output_tokens', 0),
            latency_ms=latency,
            status=status,
            error_message=error_message,
            cost_usd=self.attrs.get('cost.usd', 0.0),
            timestamp=time.strftime('%Y-%m-%dT%H:%M:%S.') + f'{int((time.time() % 1) * 1000):03d}Z',
            attributes=self.attrs,
        )
        return self.data

    def is_ended(self) -> bool:
        return self._ended
